### PR TITLE
Isolated Galaxy Sim Updates

### PR DESCRIPTION
**Pull request by clairekope**
Code additions used for [my recent paper ](https://arxiv.org/abs/2206.04706) that originally started as a fork from Devin Silvia back on Bitbucket (and is now lost). Documentation has been updated for the relevant parameters.

**GalaxySimulation Problem Type & DiskGravity External Potential**

- Multiple options for CGM structure, including rotation. One of these profiles requires the use of Grackle to calculate the cooling function on arbitrary data; this is encapsulated in GrackleCustomCoolRate
- Ability to initialize chemistry fields closer to equilibrium using precomputed pyGrackle tables. A sample script is included in the input directory. As noted in the docs, copying this feature to a different problem type may require manually overriding mu in ComputeTemperatureField, or the initial temperature will not be correctly calculated when saved to disk.
- DiskGravity stellar potentials (disk & bulge) are now accounted for when calculating the disk's circular velocity
- The DiskGravity external potential now includes an NFW potential (PointSourceGravity and DiskGravity are mutually exclusive within the isolated galaxy problem type)
- If using the DiskGravity potential, NFW halo parameters (mass & concentration) are automatically copied over to the GalaxySimulation counterparts to ensure consistency

**Star Particle Creation & Feedback**
- Added self-bound criterion and ability to use H2 mass to creation method 0 (Cen & Ostriker) as found in Hopkins et al. 2013
- Added parameters for toggling individual star formation criteria in creation method 0
- Can adjust temperature threshold used in creation method 0
- Can ramp the efficiency of thermal feedback a la minimum particle mass

### CODE REVIEW

File: src/enzo/Grid_GrackleCustomCoolRate.C [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+ 
+int CosmologyComputeExpansionFactor(FLOAT time, FLOAT *a, FLOAT *dadt);
+int GetUnits(float *DensityUnits, float *LengthUnits,
+	     float *TemperatureUnits, float *TimeUnits,
+	     float *VelocityUnits, FLOAT Time);
+int RadiationFieldCalculateRates(FLOAT Time);
+int FindField(int field, int farray[], int numfields);
+
+int grid::GrackleCustomCoolRate(int rank, int *dim, float *cool_rate,
+				float *dens, float *thrmeng,
```
Comment (jwise77) [inline src/enzo/Grid_GrackleCustomCoolRate.C:44]: See comment in Grid.h about `float` vs `FLOAT`

**Review by jwise77 — CHANGES_REQUESTED** [review]
Nice work and thanks for submitting this PR! This is a substantial addition to Enzo, and please see my inline comments that I'd like addressed before approving the PR.

File: doc/manual/source/parameters/index.rst [hunk 2]
```diff
@@ -4828,41 +4895,162 @@ Isolated Galaxy Evolution (31)
...
+    closer to equilibrium by using a pre-computed table. 
+    An example script for creating this table, which uses
+    Grackle, is ``input/gen_equilibrium_table.py``. 
+    _WARNING_ if you wish to use similar machinery for a different problem type,
+    a manual correction to the mean molecular mass may be needed within ``ComputeTemperatureField``
+    or the initial temperature may not write to disk correctly.
+    Default: 0
+``GalaxySimulationEquilibriumFile`` (external)
+    Path to the pre-computed table of equilibrium chemical species.
+    Default: none
```
Comment (jwise77) [inline doc/manual/source/parameters/index.rst:5044]: From the source, it looks like the default is "equilibrium_table_50.h5"

**Comment by clairekope** [comment]
Thanks John! I got so used to references to float being altered to double I didn't think about how if `precision-32` was used that would alter things. I will also go ahead and merge the test suite changes.

File: src/enzo/Grid.h [hunk 3]
```diff
@@ -2304,28 +2318,45 @@ int zEulerSweep(int j, int NumberOfSubgrids, fluxes *SubgridFluxes[],
 
 
   /* Initialization for isolated galaxy sims */
+  int _GalaxySimulationInitialization = 0;
   int GalaxySimulationInitializeGrid(
 				     FLOAT DiskRadius,
-				     float GalaxyMass,
-				     float GasMass,
+				     FLOAT GalaxyMass,
```
[the remark this answers is not in the document]
Comment (clairekope) [reply src/enzo/Grid.h:2324]: I thought using `FLOAT` was best practice! If it's reserved for positions and times I'll switch things to `double`.

**Comment by clairekope** [comment]
Something with my changes to `StarParticleCreation` method 1 seems to be causing a seg fault for `amr_nested_cosmology` when `opt-high` is set, but runs fine when `opt-debug` is used.

**Comment by jwise77** [comment]
> Something with my changes to `StarParticleCreation` method 1 seems to be causing a seg fault for `amr_nested_cosmology` when `opt-high` is set, but runs fine when `opt-debug` is used.

Ooof, that's the worst. I hope you can figure how what's causing the seg fault with `-O2 -g` and core dumps / gdb.

File: src/enzo/Grid_ComputeAccelerationFieldExternal.C [hunk 4]
```diff
@@ -408,57 +414,58 @@ int grid::ComputeAccelerationFieldExternal()
...
+            zheight *= LengthUnits;
+
+            radius = sqrt(rsquared) * LengthUnits;
+            rcyl = sqrt(xdisk*xdisk + ydisk*ydisk + zdisk*zdisk) * LengthUnits;
+            
+            accelsph = GravConst * 4.0*pi*rho_0*POW(Rs,3.0)
+                        *(log((Rs+radius)/Rs) - radius/(Rs+radius));
+            accelsph += (GravConst)*MBulge/POW(radius+rBulge,2);
+
+            accelcylR = GravConst*MSDisk*rcyl/sqrt(POW(POW(rcyl,2) // was rcyl converted to cgs? Yes; line 433
```
Comment (bwoshea) [inline src/enzo/Grid_ComputeAccelerationFieldExternal.C:442]: the comment seems to be inaccurate; it looks like it's now line 436 (according to GitHub at least).  Maybe just say "rcyl already converted to CGS"

File: src/enzo/Grid_ComputeAccelerationFieldExternal.C [hunk 5]
```diff
@@ -408,57 +414,58 @@ int grid::ComputeAccelerationFieldExternal()
...
+                      + POW(SDiskScaleHeightR+sqrt(POW(zheight,2) // are xpos etc already converted?
```
Comment (bwoshea) [inline src/enzo/Grid_ComputeAccelerationFieldExternal.C:443]: is this comment necessary?  raises questions...

File: src/enzo/Grid_GrackleCustomCoolRate.C [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+  g_grid_start = new Eint32[3];
+  g_grid_end = new Eint32[3];
+
+  // Fortran code will act as if there are 3 dimensions regardless
+  for (i = 0; i < rank; i++) {
+    g_grid_dimension[i] = (Eint32) dim[i];
+    g_grid_start[i] = (Eint32) 0;
+    g_grid_end[i] = (Eint32) dim[i]-1;
+  }
+  for (i = rank; i < 3; i++){
```
Comment (bwoshea) [inline src/enzo/Grid_GrackleCustomCoolRate.C:114]: If you zero out g_grid_dimension, g_grid_start, and g_grid_end in this for loop (starting at line 114), why is it also set immediately above in line 109?  Am I missing something here?

File: src/enzo/SetDefaultGlobalValues.C [hunk 7]
```diff
@@ -393,8 +393,8 @@ int SetDefaultGlobalValues(TopGridData &MetaData)
   DiskGravityStellarDiskScaleHeightz = 2.5E-4;      // Mpc
   DiskGravityStellarBulgeMass        = 1.0E10;      // Solar Masses
   DiskGravityStellarBulgeR           = 4.0E-4;      // Mpc
-  DiskGravityDarkMatterR             = 2.3E-2;      // Mpc
-  DiskGravityDarkMatterDensity       = 3.81323E-25; // CGS
+  DiskGravityDarkMatterMass          = 1e12;        // Solar Masses
```
Comment (bwoshea) [inline src/enzo/SetDefaultGlobalValues.C:396]: being super pedantic, this should probably be 1.0e12

**Comment by bwoshea** [comment]
@clairekope I made a couple of minor comments, but overall it looks fine to me!  Also, it's worth noting that the failed tests are due to a yt problem (issues getting ICs from the YT data hub), not a crash due to Enzo itself, so I'm comfortable merging without those tests passing.

File: src/enzo/Grid_ComputeAccelerationFieldExternal.C (hunk 5 above)
Comment (clairekope) [reply src/enzo/Grid_ComputeAccelerationFieldExternal.C:443]: These comments were largely for my own tracking of what was going on with units. I've deleted them and adjusted a preceding comment for clarification

File: src/enzo/Grid_GrackleCustomCoolRate.C (hunk 6 above)
Comment (clairekope) [reply src/enzo/Grid_GrackleCustomCoolRate.C:114]: The two loops (line 109 and line 114) have different start and end values. The Fortran code called to calculate the cooling time will _always_ act as if the problem is three dimensional. So if you're calling this function for a 1D problem, the second loop sets entries corresponding to the unused second and third dimensions to zero. This is maybe a weird way of going about this, but I wanted this function to be written generally. I've updated the comments to hopefully make this clearer.

Comment (bwoshea) [reply src/enzo/Grid_GrackleCustomCoolRate.C:116]: whoops, I totally missed that.  Thank you!

**Comment by bwoshea** [comment]
@clairekope the YT devs fixed the issues on their side, so CircleCI is running now.  Assuming the tests pass, I'll approve and merge!

**Comment by bwoshea** [comment]
All tests have passed.  I'm going to approve and merge!

**Review by bwoshea — APPROVED** [review]

**Pull request merged**